bots/doc.go

```diff
@@ -0,0 +1,139 @@
+with its designated third-party service, and the requests must be sent by
+an authorized CONIKS client via the service's protocol to the proxy's
+reserved account. As a result, a proxy trusts that any requests sent
+by CONIKS clients to its account originated from uncompromised
+third-party accounts.
+
+CONIKS servers: bots proxies relay verified CONIKS registration requests
+to a designated CONIKS server. The server and proxies may run as separate
+processes on the same machine, or on separate machines, depending on the
+communication service's infrastructure. For authentication, bots proxies
+can be configured to digitally sign all requests they relay, or to
+establish a secure network connection with the server. The server then
+only accepts incoming registration requests originating from authenticated
+account proxies.
+
+Privacy
+
+bots seeks to provide the same strong privacy guarantees as CONIKS.
+CONIKS clients using bots only store the generated key pairs and
+registered third-party usernames. bots only sends the username and
+the public key as an opaque data blob to the corresponding account proxy.
+bots also ensures that the clients do not gain access to any other
+third-party account information.
+
+The account proxies only have access to a user's username with their
+designated service and the received public key, and this is
+the only user information they send to the server.
+Proxies cannot access any other account information.
+
+Third-Party Account Verification Protocol
+
+1. The user registers her third-party account with her communication
+client, thereby authorizing the client and the integrated CONIKS client
+to connect to her third-party account.
+
+2. The CONIKS client generates a key pair for the new user account
+and stores the public and private key locally.
+
+3. On behalf of the connected account, the client sends a registration
+request
+    req = (username, opaque public key data blob)
+to the designated account proxy using the service's communication
+protocol.
+
+4. The proxy receives the request at its reserved account via
+the service's protocol, and verifies that the sender
+is an authorized client.
+
+5. The proxy sends the verified registration request to its designated
+CONIKS server.
+
+6. Once the CONIKS server receives the request and authenticates the proxy,
+it registers the username-to-public key mapping in its CONIKS directory.
+The server sends a cryptographic proof of registration to the proxy,
+which relays the response back to the CONIKS client.
+
+Supported Third-Party Services
+
+The current implementation supports third-party account verification of
+Twitter accounts. The account proxy connects to its reserved Twitter
+account using OAuth, and it only accepts CONIKS registration
+requests received as Twitter direct messages sent from CONIKS clients
+that have been authorized by a legitimate Twitter account.
+
+Limitations
```

Comment:
> Hmm, it seems like we're flipping the first party and third party.

Sounds like it, sorry.  In my comment I meant,

- First party: keyserver is run by identity provider
- Third party: keyserver is run by communication service (or really anyone) on behalf of the identify provider

> I'm assuming, you're referring to the "proxy cannot verify the ownership of the key" limitation here.

No, verifying the ownership of the key is limited by the `allowsPublicVisibility` flag, and seems the same in the two cases.

I was talking about, "we avoid exposing the user's public key to the registration proxy" (which is similarly affected by that flag, but let's ignore that for now).  Is that even true though, we aren't encrypting the data?  What I meant above is that since the proxy and keyserver are both currently being run by the same person, it's not really an additional limitation.

> This is definitely true, but I'm a bit confused as to how this issue relates to the registration bot.

I guess because this is where the breakdown happens?  In the third-party case (my terminology above), we have no proof that the identity provider requested the registration.  If we did, then we'd be able to determine who was being malicious.  But also because I wasn't sure where else we had planned to write this up.